ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,476 @@
+    Future<String> future = router.putBlob(opInfo.properties, opInfo.userMetadata, putChannel, callback);
+    TestFuture<String> testFuture = new TestFuture<String>(future, genLabel("putBlob", nonBlocking, false), opInfo) {
+      @Override
+      void check() {
+        checkBlobId(get(), getOperationName());
+      }
+    };
+    opInfo.futures.add(testFuture);
+  }
+
+  private void startGetBlobInfo(boolean nonBlocking, final boolean afterDelete, final OperationInfo opInfo) {
+    Callback<BlobInfo> callback = new TestCallback<>(opInfo, afterDelete);
+    Router router = nonBlocking ? nonBlockingRouter : coordinatorBackedRouter;
+    Future<BlobInfo> future = router.getBlobInfo(opInfo.blobId, callback);
+    TestFuture<BlobInfo> testFuture =
+        new TestFuture<BlobInfo>(future, genLabel("getBlobInfo", nonBlocking, afterDelete), opInfo) {
+          @Override
+          void check() {
+            if (afterDelete) {
+              try {
+                future.get();
+                Assert.fail("Blob should have been deleted in operation: " + getOperationName());
+              } catch (Exception ignored) {
+              }
+            } else {
+              checkBlobInfo(get(), opInfo, getOperationName());
+            }
+          }
+        };
+    opInfo.futures.add(testFuture);
+  }
+
+  private void startGetBlob(boolean nonBlocking, final boolean afterDelete, final OperationInfo opInfo) {
+    Callback<ReadableStreamChannel> callback = new TestCallback<>(opInfo, afterDelete);
+    Router router = nonBlocking ? nonBlockingRouter : coordinatorBackedRouter;
+    Future<ReadableStreamChannel> future = router.getBlob(opInfo.blobId, callback);
+    TestFuture<ReadableStreamChannel> testFuture =
+        new TestFuture<ReadableStreamChannel>(future, genLabel("getBlob", nonBlocking, afterDelete), opInfo) {
+          @Override
+          void check() {
+            if (afterDelete) {
+              try {
+                future.get();
+                Assert.fail("Blob should have been deleted in operation: " + getOperationName());
+              } catch (Exception ignored) {
+              }
+            } else {
+              checkBlob(get(), opInfo, getOperationName());
+            }
+          }
+        };
+    opInfo.futures.add(testFuture);
+  }
+
+  private void startDeleteBlob(boolean nonBlocking, final OperationInfo opInfo) {
+    Callback<Void> callback = new TestCallback<>(opInfo, false);
+    Router router = nonBlocking ? nonBlockingRouter : coordinatorBackedRouter;
+    Future<Void> future = router.deleteBlob(opInfo.blobId, callback);
+    TestFuture<Void> testFuture = new TestFuture<Void>(future, genLabel("deleteBlob", nonBlocking, false), opInfo) {
+      @Override
+      void check() {
+        get();
+      }
+    };
+    opInfo.futures.add(testFuture);
+  }
+
+  private void startAwaitCreation(final OperationInfo opInfo) {
+    notificationSystem.awaitBlobCreations(opInfo.blobId);
```

Comment:
The `awaitBlobCreations()` method has a specific timeout (which is 2 second for now), and this could potentially makes the test to run unfavorably long time. I do not see a good way to fix, but just raise a flag here.